src/TabBar.tsx

```diff
@@ -1,130 +1,142 @@
 import {
-  View,
-  TouchableOpacity,
   Dimensions,
-  ScrollView,
   LayoutChangeEvent,
   LayoutRectangle,
-  SectionListData,
   RegisteredStyle,
-  ViewStyle
-} from 'react-native';
-const WindowWidth = Dimensions.get('window').width;
+  ScrollView,
+  SectionListData,
+  StyleSheet,
+  TouchableOpacity,
+  View,
+  ViewStyle,
+} from "react-native"
+
+const windowWidth = Dimensions.get("window").width
 
 interface IProps {
-  sections: SectionListData<any>[];
-  renderTab: (section: SectionListData<any>) => React.ReactNode;
-  tabBarStyle?: ViewStyle | RegisteredStyle<ViewStyle>;
-  currentIndex: number;
-  onPress: (index: number) => void;
+  currentIndex: number
+  onPress: (index: number) => void
+  sections: Array<SectionListData<any>>
+  tabBarStyle?: ViewStyle | RegisteredStyle<ViewStyle>
+  renderTab: (section: SectionListData<any>) => ReactNode
 }
 
 interface ITabMeasurements {
-  left: number;
-  right: number;
-  width: number;
-  height: number;
+  left: number
+  right: number
+  width: number
+  height: number
 }
 
 interface ITabsLayoutRectangle {
-  [index: number]: ITabMeasurements;
+  [index: number]: ITabMeasurements
 }
 
-export default class TabBar extends React.PureComponent<IProps, any> {
-  private scrollView: React.RefObject<ScrollView> = React.createRef();
-  private _tabContainerMeasurements!: LayoutRectangle;
-  private _tabsMeasurements: ITabsLayoutRectangle = {};
-
-  componentDidUpdate(prevProps: IProps) {
-    if (this.props.currentIndex !== prevProps.currentIndex) {
-      if (this.scrollView.current) {
-        this.scrollView.current.scrollTo({
-          x: this.getScrollAmount(),
-          animated: true
-        });
+const TabBar: FC<IProps> = ({ sections, tabBarStyle, currentIndex, renderTab, onPress }) => {
+  const scrollViewRef = useRef<ScrollView>()
+  const localIndex = useRef(0)
+  const [tabContainerMeasurements, setTabContainerMeasurements] = useState<LayoutRectangle>()
+  const [tabsMeasurements, setTabsMeasurements] = useState<ITabsLayoutRectangle>({})
+
+  const s = useMemo(
+    () => ({
+      view: [
+        {
+          width: windowWidth,
+        },
+        tabBarStyle,
+      ],
+    }),
+    [tabBarStyle],
+  )
+
+  // default method
+  const getScrollAmount = useCallback(() => {
+    if (!tabContainerMeasurements) {
+      return
+    }
+    const pageOffset = 0
+    const position = currentIndex
+    const containerWidth = windowWidth
+    const tabWidth = tabsMeasurements[position].width
+    const nextTabMeasurements = tabsMeasurements[position + 1]
+    const nextTabWidth = (nextTabMeasurements && nextTabMeasurements.width) || 0
+    const tabOffset = tabsMeasurements[position].left
+    const absolutePageOffset = pageOffset * tabWidth
+    let newScrollX = tabOffset + absolutePageOffset
+    newScrollX -= (containerWidth - (1 - pageOffset) * tabWidth - pageOffset * nextTabWidth) / 2
+    newScrollX = newScrollX >= 0 ? newScrollX : 0
```

Comment:
This operation can be inlined, but keep it this way if it's easier for you to read.